### PR TITLE
Add bwc layer for 'romanian' analyzer

### PR DESCRIPTION
**Pull request by cbuescher**
The 'romanian' language analyzer has been improved in Lucene 10 in two important ways. First, the snowball stemmer has been modified to work with s-comma and t-comma characters but only with their cedilla forms used when Romanian didn't have full Unicode support (https://github.com/snowballstem/snowball/pull/177). Second, the analyzer now contains a normalization step to map cedilla forms to forms with comma.

In order to maintain backwards compatibility with existing indices, this change moves the Lucene 9 stemmer over to the analysis module was a deprecated variant and creates the analyzer for existing indices with the "old" stemmer and without the normalization step. New indices automatically run with the improved behaviour.

### CODE REVIEW

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/LegacyRomanianStemmer.java [hunk 1]
```diff
@@ -0,0 +1,744 @@
...
+
+package org.elasticsearch.analysis.common;
+
+import org.tartarus.snowball.Among;
+
+/**
+* This class implements the stemming algorithm defined by a snowball script.
+* NOTE: This is the RomanianStemmer used in Lucene 9
+*
+* @deprecated this class exists for backwards compatibility of indices created with Lucene 9
```
Comment (benwtrent) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/LegacyRomanianStemmer.java:16]: I don't think we should do this. We aren't providing bwc layer for ANY other stemmer, its weird that we are for this one.

File: muted-tests.yml [hunk 2]
```diff
@@ -339,6 +339,9 @@ tests:
 - class: org.elasticsearch.threadpool.SimpleThreadPoolIT
   method: testThreadPoolMetrics
   issue: https://github.com/elastic/elasticsearch/issues/108320
+- class: org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT
```
Comment (benwtrent) [inline muted-tests.yml:342]: did you mute this in main already? We should make sure this is actually muted in the appropriate PR and then update this branch once that mute is merged.

**Review by benwtrent — COMMENTED** [review]
I am unsure about providing the legacy stemmer, we aren't doing this for any other language, we shouldn't do it for this one. 

Though, I am good with the bwc analyzer that skips the normalization process on old indices.

File: muted-tests.yml (hunk 2 above)
Comment (cbuescher) [reply muted-tests.yml:342]: This is going away with the next merge with the lucene_snapshot branch, I just cherry-picked it to pass CI

**Comment by cbuescher** [comment]
> I am unsure about providing the legacy stemmer, we aren't doing this for any other language

I understand, the problem here is that the updated RomanianStemmer also does normalization of the cedilla to comma forms. So even without the normalizer we wouldn't be able to reproduce the right output for legacy indices, leading to a large mismatch tokens from the search-analyzer to the indexed ones, lowering recall.
As far as I understand the use of the legacy forms is not as common as the new ones but not negligible if I look at [this analysis done by Robert](https://github.com/apache/lucene/pull/12172#issuecomment-1452764003)

**Comment by benwtrent** [comment]
> I understand, the problem here is that the updated RomanianStemmer also does normalization of the [cedilla to comma forms](https://en.wikipedia.org/wiki/Romanian_alphabet#Comma-below_(%C8%99_and_%C8%9B)_versus_cedilla_(%C5%9F_and_%C5%A3)). So even without the normalizer we wouldn't be able to reproduce the right output for legacy indices, leading to a large mismatch tokens from the search-analyzer to the indexed ones, lowering recall....Its hard to even write a tests where the input term has the "old" cedilla form of a character and the custom analyzer without the normalization step leaves that untouched. The new Stemmer version always seems to do the normalization by itself.


OK, I originally mis-read the change and it seemed like somebody could provide access to the legacy stemmer, but instead we just dynamically do the correct thing on the old indices.

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/LegacyRomanianStemmer.java [hunk 3]
```diff
@@ -0,0 +1,744 @@
...
+*/
+@Deprecated
+@SuppressWarnings("checkstyle:DescendantToken")
+public class LegacyRomanianStemmer extends org.tartarus.snowball.SnowballStemmer {
+
```
Comment (benwtrent) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/LegacyRomanianStemmer.java:21]: We should remove `Deprecated` and make this class package private.

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/RomanianAnalyzerProvider.java [hunk 4]
```diff
@@ -9,28 +9,60 @@
...
+            analyzer = new StopwordAnalyzerBase(Analysis.parseStopWords(env, settings, RomanianAnalyzer.getDefaultStopSet())) {
+
+                protected Analyzer.TokenStreamComponents createComponents(String fieldName) {
+                    final Tokenizer source = new StandardTokenizer();
+                    TokenStream result = new LowerCaseFilter(source);
+                    result = new StopFilter(result, stopwords);
+                    if (stemExclusionSet.isEmpty() == false) {
+                        result = new SetKeywordMarkerFilter(result, stemExclusionSet);
+                    }
+                    result = new SnowballFilter(result, new LegacyRomanianStemmer());
```
Comment (benwtrent) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/RomanianAnalyzerProvider.java:52]: I understand now, we can keep the legacy stemmer, but we should make sure this doesn't write a deprecation warning as we are dynamically doing the correct thing based on index versions.

**Review by benwtrent — COMMENTED** [review]
I understand more clearly now.

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/LegacyRomanianStemmer.java (hunk 3 above)
Comment (cbuescher) [reply modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/LegacyRomanianStemmer.java:21]: Sure np

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/RomanianAnalyzerProvider.java (hunk 4 above)
Comment (cbuescher) [reply modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/RomanianAnalyzerProvider.java:52]: correct

**Comment by cbuescher** [comment]
@benwtrent I pushed changes addressing your comments

**Review by benwtrent — APPROVED** [review]

**Pull request merged**